pkg/build/strategies/sti/sti.go

```diff
@@ -410,8 +406,28 @@ func (b *STI) Execute(command string, config *api.Config) error {
 		Env:             buildEnv,
 		PostExec:        b.postExecutor,
 	}
+
 	if !config.LayeredBuild {
-		opts.Stdin = tarFile
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		uploadDir := filepath.Join(config.WorkingDir, "upload")
+
+		// TODO: be able to pass a stream directly to the Docker build to avoid the double temp hit
+		r, w := io.Pipe()
+		go func() {
+			var err error
+			defer func() {
+				w.CloseWithError(err)
+				if r := recover(); r != nil {
+					glog.Errorf("recovered panic: %#v", r)
+				}
+				wg.Done()
+			}()
+			err = b.tar.CreateTarStream(uploadDir, false, w)
+		}()
+
+		opts.Stdin = r
```

Comment:
is there a reason you can't move all this logic (setting up the stdin stream) above the creation of the RunContainerOptions struct, and populate the opts.Stdin field as part of creating that struct, instead of doing it here?